### PR TITLE
Improve faulty parsing

### PR DESCRIPTION
**Pull request by privat**
Some improvements (and tests) related to the parsing of faulty code.

Basically, these are test cases on faulty code and the required fixes to the tests become green.
This PR (and future ones) focuses on having usable and consistent AST when parsing broken code.

Note: since I hacked some low level image stuff, I hope I did not break too many tests :)

### CODE REVIEW

**Comment by jecisc** [comment]
A failure is related.

It's not the tests breaking but the building 

```
Returning an object but pool is fullSyntax Error on line 1: 'Missing opener for closer: )'

======================================================

1: (RBParenthesesErrorNode content: {RBMessageNode receiver: (RBParseErrorNode errorMessage: 'Variable or expression expected' value: '' at: 3) selector: #+ arguments: {RBParseErrorNode errorMessage: 'Variable or expression expected' value: '' at: 4. }. } start: 1 stop: 3 errorMessage: ''')'' expected') value: '(') valueAfter: ''

                                                                                                                                                                                                                                                                                                                          _^_

```

**Comment by privat** [comment]
The building seems fine, it is the tests that are badly broken, as the whole test phase is just aborted.
The log, is mostly garbage that tests execution are printing (it could be cleaned up or replaced by more useful log information).  The scary string `Returning an object but pool is full` is unrelated to the breakage, as it appears the same is success CI runs. The red part is new but not really useful :(

I tried to reproduce, and the culprit seems `RBDumpVisitorTest>>#testErrorNodeDump` that is consistent with the red message in the CI logs and related to some of the change made by the PR. But running it locally broke and froze my image (it could explain the abort behavior in the CI).

Anyway, this will be a fun debugging morning, I guess :shrug:

**Comment by privat** [comment]
Some tests are still failing, I'm on it. But people can review anyway :)

I improved the method RBFormatterTest>>#testParseError to handle a lot more of faulty expression. It allowed me to find more case that need fixing.
See https://github.com/privat/pharo/blob/improve-faulty-parsing/src/AST-Core-Tests/RBFormatterTest.class.st#L74 if the diff is not nice enough to read.

If you have suggestion for more broken expressions, I will gladly take them (@MarcusDenker ?)

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/privat/pharo">pharo/RBFormatterTest.class.st at improve-faulty-parsing · privat/pharo</a></strong></div><div>Pharo is a dynamic reflective pure object-oriented language supporting live programming inspired by Smalltalk. - pharo/RBFormatterTest.class.st at improve-faulty-parsing · privat/pharo</div></blockquote>

**Comment by privat** [comment]
Rhaaa. merge conflicts are related to the numerous cleanup PR :(

**Comment by privat** [comment]
merge conflicts solved, and all tests are green (except for the Windows box, I assume this is unrelated)

File: src/AST-Core/RBParser.class.st [hunk 1]
```diff
@@ -1003,7 +1007,13 @@ RBParser >> parseStatementInto: statementList periodList: periods withAcceptedSt
 		ifFalse: [ node := self parseAssignment ].
 
 	"If the parsing failed because it did not match anything, consume the current token and continue parsing."
-	startOfStatementToken = currentToken ifTrue: [ self step ].
+	startOfStatementToken = currentToken ifTrue: [
+		"We did not progress, its mean that 1. we have a error node and 2. we are missing something (or found something unexpected, its the same thing)."
+		node isParseError ifFalse: [ 1 haltIfTest ].
```
Comment (MarcusDenker) [inline src/AST-Core/RBParser.class.st]: There are some halts here

**Comment by privat** [comment]
Nice catch, I let the halts to detect unexpected code path when running the full CI, but I forgot to remove them.

**Pull request merged**